clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1176,18 +1176,25 @@ private ClusterAndWaitTime waitOnMetadata(String topic, Integer partition, long
                 metadata.awaitUpdate(version, remainingWaitMs);
             } catch (TimeoutException ex) {
                 // Rethrow with original maxWaitMs to prevent logging exception with remainingWaitMs
-                throw new TimeoutException(
-                        String.format("Topic %s not present in metadata after %d ms.",
-                                topic, maxWaitMs));
+                final String errorMessage = String.format("Topic %s not present in metadata after %d ms.",
+                        topic, maxWaitMs);
+                if (metadata.getError(topic) != null) {
```

Comment:
I may be missing something, but won't this always be false in practice? (and we'll never be able to get a root cause here). This is my reasoning: if we got TimeoutException while awaiting a metadata update, it means that the metadata version did not get bumped (`metadata.update` never happened), therefore, we never made it to the point where we get the errors from the response [here](https://github.com/apache/kafka/blob/a27a458d03a173691053c21ebcfb21f11d3cad12/clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetadata.java#L135). Makes sense? 

Honestly, at this point I don't see we're able to get any cause for the timeout (we simply haven't received a response for the MetadataRequest, which is what should trigger the update [here](https://github.com/apache/kafka/blob/2b491f9611cb8514687b4030f3da95dc7573e1da/clients/src/main/java/org/apache/kafka/clients/NetworkClient.java#L1220))

Reply:
Thanks. It makes sense! The `if` clause always returns `false`.

Reply:
Did we address this comment?

Reply:
Also to be clear, there are two types of timeouts here 
1 -- we never get a response back
2 -- we are getting retriable errors over and over (ie unknown topic or partition) until the timeout.

Case 2 is what we hope to improve.

Reply:
Good point, I had 1 in mind when not seeing a cause to add here, but defitenly 2 may happen if we are retrying continuously and time runs out while awaitingUpdate (and not on the following elapsed check). So we would want to throw with the cause if errors not null as you initially had @aliehsaeedii , makes sense to me.  

Note that here I don't think we need any check on retriables I would say, because if metadata.getError had had any fatal, it would have been thrown before hitting this line (because the `maybeThrowExceptionForTopic` happens right before this, in the previous iteration, and timeout here means no other metadata update happened). Makes sense?

Reply:
Yes, makes sense to me.

Reply:
I have a couple questions:
1. Looks like we're just changing the TimeoutException handling on this codepath, but wouldn't we want to change this across the board, i.e. whenever we generate TimeoutException we actually add the last retriable error (unless it's a true timeout and we simply didn't get a reply)?
2. Rather than do this error state tunnelling across multiple levels, wouldn't it be simpler to just add the root cause whenever we decide to initially throw TimeoutException (i.e. when we get a retriable error but throw TimeoutException instead)?

Also I'm not sure if we get an error in the metadata request if the topic is just missing -- looks the metadata requests would be successful, just not contain the topic.

Reply:
> Also I'm not sure if we get an error in the metadata request if the topic is just missing -- looks the metadata requests would be successful, just not contain the topic.

Yes, this was something I raised internally as well. I think if the error message is confessing that the topic is missing, it is missing even with no error.

Reply:
I will let @aliehsaeedii jump in here, but my 2 cents regarding @artemlivshits point 1 and 2. I totally agree with the view on both, but seems to me that's what we're getting here already?? Trick is that we have different timeouts mixed up here, but let's clarify the concepts: the timeout we want to add a cause to is a "global timeout on retries" (not timeout on a single awaitUpdate). With this, we are indeed :

1. "whenever we generate TimeoutException we actually add the last retriable error" -> we add the cause if we have a last retriable, on the 2 points where a new "global timeout on retries" is created
2. "add the root cause whenever we decide to initially throw TimeoutException" -> similar as above, on the 2 points where we check if the global timeout is expired, we create the exception with the cause (if we have one from a previous retry)

Makes sense?

Reply:
> wouldn't it be simpler to just add the root cause whenever we decide to initially throw TimeoutException

The 1st timeout exception originated from [here](https://github.com/apache/kafka/blob/645eefe95aa764fdcd1c6390bb2a2e8dc14cde84/clients/src/main/java/org/apache/kafka/common/utils/SystemTime.java#L58) and we decide on throwing the 2nd one if `elapsed >= maxWaitMs`. Therefore, to have the root cause, we needed kanda tunnelling across multiple levels.

Reply:
> , but seems to me that's what we're getting here already??

I presume there are other timeouts on other code paths, not just in getting metadata?  E.g. when we produce to a topic and the leader has stale metadata, so the client sees the topic present, but the leader hasn't caught up on latest metadata and says that it doesn't know about this topic.  commitTransaction is another codepath that returns timeouts because it waits for sends to complete and sends may be failing with retriable error.  Etc. 

> the timeout we want to add a cause to is a "global timeout on retries" (not timeout on a single awaitUpdate)

I understand that, the intent is to translate the timeout message, not generate a new one, so it can be structured as

```
  String errorMessage = /* translated message with total timeout */
  throw TimeoutException(errorMessage, ex.getCause());
```

Then we can still follow the rule "whoever is creating TimeoutException needs to set the cause" and pass this info more locally without the need to save and tunnel errors around.

Reply:
That's a good point. I would prefer to do some follow up PR to unify all this code and clean it up nicely. I can file a JIRA for tracking and we should tackle it as immediate follow up before 3.9 goes out.

Reply:
sounds good

Reply:
Thanks. Filed https://issues.apache.org/jira/browse/KAFKA-17019